src/software/backend/input/replay_logging/replay_test.cpp

```diff
@@ -0,0 +1,101 @@
+#include <google/protobuf/util/message_differencer.h>
+#include <gtest/gtest.h>
+
+#include "software/backend/input/replay_logging/replay_logger.h"
+#include "software/backend/input/replay_logging/replay_reader.h"
+#include "software/multithreading/subject.h"
+
+// the working directory of tests are the bazel WORKSPACE root (in this case, src)
+// this path is relative to the current working directory, i.e. the bazel root
+constexpr const char* REPLAY_TEST_PATH_SUFFIX =
+    "software/backend/input/replay_logging/test_replay";
+
+namespace fs = std::experimental::filesystem;
+
+// hahaha test subject because it's a subject but also a test subject
+class TestSubject : public Subject<SensorMsg>
+{
+   public:
+    void sendValue(SensorMsg val)
+    {
+        sendValueToObservers(val);
+    }
+};
+
+TEST(ReplayTest, test_read_and_write_replay)
+{
+    // unfortunately due to the unavailablity of ordering tests and functions that
+    // generate test files that actually work, we have to read recorded replays back,
```

Comment:
Can we not write files that "actually work"? Isn't that what the logger's doing?

Reply:
I don't know what exactly was going wrong, but with manually created test messages, there was an issue where we'd always have a pure virtual function call (same issue brought up in chat before @akhilveeraghanta), but there was no `required bool` anywhere in the 'tree' of both `SSL_WrapperPacket` and `Referee` if i remember correctly. This is probably worth digging into deeper in the future.